xcm/src/v2/mod.rs

```diff
@@ -0,0 +1,705 @@
+// Copyright 2020 Parity Technologies (UK) Ltd.
+// This file is part of Cumulus.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Cumulus.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Version 1 of the Cross-Consensus Message format data structures.
+
+use super::v1::{Order as OldOrder, Response as OldResponse, Xcm as OldXcm};
+use crate::DoubleEncoded;
+use alloc::{vec, vec::Vec};
+use core::{
+	convert::{TryFrom, TryInto},
+	fmt::Debug,
+	result,
+};
+use derivative::Derivative;
+use parity_scale_codec::{self, Decode, Encode};
+
+mod traits;
+
+pub use traits::{Error, ExecuteXcm, Outcome, Result, SendError, SendResult, SendXcm};
+// These parts of XCM v1 have been unchanged in XCM v2, and are re-imported here.
+pub use super::v1::{
+	Ancestor, AncestorThen, AssetId, AssetInstance, BodyId, BodyPart, Fungibility,
+	InteriorMultiLocation, Junction, Junctions, MultiAsset, MultiAssetFilter, MultiAssets,
+	MultiLocation, NetworkId, OriginKind, Parent, ParentThen, WildFungibility, WildMultiAsset,
+};
+
+#[derive(Derivative, Encode, Decode)]
+#[derivative(Clone(bound = ""), Eq(bound = ""), PartialEq(bound = ""), Debug(bound = ""))]
+#[codec(encode_bound())]
+#[codec(decode_bound())]
+pub struct Xcm<Call>(pub Vec<Instruction<Call>>);
```

Comment:
I'm wondering if there's a way to enforce some kind of well-formedness with the XCMs, because some instructions don't make sense by their own without another happening prior, e.g. `WithdrawAsset` should have been executed before calling `BuyExecution`. We currently have a sort of idempotence when instructions such as `BuyExecution` is called before `WithdrawAsset` by simply failing to execute, but I wonder if we can leverage Rust's static analysis to aid in XCM authors in writing well-formed XCMs.

It's definitely going to be tricky to think about, since XCMs are not necessarily self-contained and state transitions across XCMs should be allowed (one processed XCM could've already called `WithdrawAsset`, and the next one calls `BuyExecution`).

Reply:
I think the only way is build a state machine and each instruction is an action, and for some state, some actions are not allowed.
Not sure if there is already existing tools to define this, but I image it will be cumbersome to use.
Alternately we make this a proper scripting language and have some linter to check it.

Reply:
A linter could probably identify *definitely* bad programs, but I don't think a general validator is possible.